### PR TITLE
feat: implement notification system

### PR DESCRIPTION
**Pull request by BeierKevin**
# PR description
*Describe your changes in detail here*

A implementation of a simple warning system where you just add and remove a Warning Object. When you do so it will display a little notification to the user, but only if the previous notification has either timed out or been closed by the user manually.

Some feedback would be very nice, because I wasn't sure what the functionalities should be.

This PR has been done in pair programming to gradually introduce @roman533 into the workprocess

# Definition Of Done (DoD)
*This PR can be squashed / merged if*
- [x] a developer is assigned
- [x] the PR is NOT estimated
- [x] the PR is labeled
- [x] the PR is NOT assigned to the current sprint
- [x] a meaningful title has been set according to https://www.conventionalcommits.org/
- [x] the PR is described in detail
- [x] the PR links to an issue
- [x] the PR has been reviewed

*Add additional conditions here if necessary for this PR*

Fix: #109

### CODE REVIEW

**Review by Claiyc — COMMENTED** [review]
Doesn't look bad. examples need to be removed though.

**Comment by BeierKevin** [comment]
Yes the example need to be removed, just left them in for testing and showcasing purposed

**Comment by BeierKevin** [comment]
I have now removed the test code, and commented the code a bit more

File: src/composables/useWarningSystem.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * Interface that can be used to define a warning.
+ */
+export interface Warning {
+    message: string; // the message to display
+    timeout?: number; // in milliseconds
+    color?: string; // can be hex, rgb, rgba, hsl, hsla, or something like warning, error, info, success more info: https://vuetifyjs.com/en/api/v-snackbar/#props-color
+    isMultiLine?: boolean;
+    isActive: boolean; // if the warning is active, this has to be TRUE to show the warning in the queue at all, otherwise it will be removed from the queue
```
Comment (Claiyc) [inline src/composables/useWarningSystem.ts:44]: why is this required? We will never have a warning that is not displayed, so from my pov it's sufficient to just display all warnings that exist and entirely deleting it removes it.

File: src/App.vue
[the remark this answers is not in the document]
Comment (BeierKevin) [reply src/App.vue:1]: They aren't accutally below each other, only the first is displayed.

The below each other stack, isn't nativly supported by this component, I would need to build a wrapper around the existing component but it would be possible.

[the remark this answers is not in the document]
Comment (Claiyc) [reply src/App.vue:1]: your call. If it's possible and not too much work then do it. You may also split it into a seperate issue and just leave it in the backlog for now

[the remark this answers is not in the document]
Comment (BeierKevin) [reply src/App.vue:1]: I think we can make a seperate issue out of this, because the general implementation of the system aswell as the displaying of it works, the Stacking is marely some design details which we can discuss

File: src/composables/useWarningSystem.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+};
+
+/**
+ * Interface that can be used to define a warning.
+ */
+export interface Warning {
+    message: string; // the message to display
+    timeout?: number; // in milliseconds
+    color?: string; // can be hex, rgb, rgba, hsl, hsla, or something like warning, error, info, success more info: https://vuetifyjs.com/en/api/v-snackbar/#props-color
+    isMultiLine?: boolean;
```
Comment (Claiyc) [inline src/composables/useWarningSystem.ts:43]: What about this, is this actually required? Seems like a overcomplicated step for just a warning message (I know it's a native attribute of this component, but can't we somehow determine it automatically or just default it to true?)

Comment (BeierKevin) [reply src/composables/useWarningSystem.ts:43]: I won't know how to determine it automatically, but sure we can set it to true by default

File: src/App.vue
[the remark this answers is not in the document]
Comment (BeierKevin) [reply src/App.vue:1]: #130 issue created

File: src/App.vue [hunk 3]
```diff
@@ -203,7 +203,7 @@
             v-model="showWarning"
             :timeout="warnings[0].timeout || 2000"
             @update:model-value="dismissWarning(warnings[0])"
-            :multi-line="warnings[0].isMultiLine"
+            :multi-line="true"
```
Comment (Claiyc) [inline src/App.vue:206]: did you test whether this is an issue with short or long messages or messages with linebreaks?

Comment (BeierKevin) [reply src/App.vue:206]: Does this look good enough ? 

image

Comment (BeierKevin) [reply src/App.vue:206]: A few lines below, i set the max width of the snackbar so it will never get bigger

Comment (Claiyc) [reply src/App.vue:206]: overflow scroll? Or how do you enforce a max-width?

Comment (BeierKevin) [reply src/App.vue:206]: If you text is too long then it will flow out of the application.

image

Max-width is a prop on the component

Comment (BeierKevin) [reply src/App.vue:206]: What I could do is something like adding a scrollbar: 
image

and defining the max-height of it in a seperate css class

Comment (Claiyc) [reply src/App.vue:206]: seems like a better idea, doesn't it?

Comment (BeierKevin) [reply src/App.vue:206]: Or we could use something else like after X chars its cut of wiht ...

image

Comment (BeierKevin) [reply src/App.vue:206]: But tbh when will er ever have such a Long warning message that er need a scroll bar, i think The trunking Version ja cleaner

Comment (Claiyc) [reply src/App.vue:206]: your call

Comment (BeierKevin) [reply src/App.vue:206]: I am trying to look for a solution for the stacked version but currently I am facing some issues which I try to figure out

image

Comment (Claiyc) [reply src/App.vue:206]: I'd also recommend to stick to one color (something red)

Comment (BeierKevin) [reply src/App.vue:206]: Currently it is possible to pass something in like "warning", "error", "info", "success" then it gets a specific color from it, so this warning system could be basically be renamed as a notification system because it has the possibilities to notify the user from error that Accor, when something successes and informational notifications

Comment (Claiyc) [reply src/App.vue:206]: that works for me as well

Comment (BeierKevin) [reply src/App.vue:206]: Ok, but first ill try to first the stacking of the notifications

**Review by roman533 — DISMISSED** [review]

File: src/App.vue [hunk 4]
```diff
@@ -307,24 +320,94 @@ async function regenerateAccessToken() {
...
+        color: 'warning',
+        timeout: -1,
+        isActive: true,
+    });
```
Comment (Claiyc) [inline src/App.vue:410]: remove example code

File: src/App.vue [hunk 5]
```diff
@@ -307,24 +320,94 @@ async function regenerateAccessToken() {
...
+function dismissNotification(item: Notification, index: number) {
+    useNotificationSystem().removeNotification(item);
+    calcMargin(index);
+}
+
+/**
+ * Function which will truncate the text of the notification if it is too long to display
+ * @param index The index of the notification
+ */
+function truncatedText(index: number): string {
```
Comment (Claiyc) [inline src/App.vue:351]: I don't think that this should be our final solution. Using a scrollbar would be better in this case.

**Review by Claiyc — CHANGES_REQUESTED** [review]

File: src/App.vue (hunk 5 above)
Comment (BeierKevin) [reply src/App.vue:351]: The current stacking of the notifications doesn't work if the notifications have different sizes thats why i truncated the text

File: src/App.vue (hunk 4 above)
Comment (BeierKevin) [reply src/App.vue:410]: Will do after we resolved other issues

File: src/App.vue (hunk 5 above)
Comment (Claiyc) [reply src/App.vue:351]: I understand that, however the intermediate solution should not be to cut the warning message either.

It seems like putting them below one another does not work at all. Pressing the snackbar button twice results in them overlapping

Comment (BeierKevin) [reply src/App.vue:351]: Then you haven't pulled the lates changes i think because for me they aren't stacking anymore

Comment (Claiyc) [reply src/App.vue:351]: ok it doesn't stack but it only shows 4 messages at max at a time. And it is not clear whether there are more messages below. Can you maybe add a scroll-functionality there, so that you're able to scroll through the list of messages? Also maybe cut the last message half-way so that it's clear that there is another message below.

Actually I don't see the necessity to limit the number of messages at all. The user is SUPPOSED to acknowledge the message by closing it, so having an ugly mass of messages on the screen is not an issue imo

Comment (BeierKevin) [reply src/App.vue:351]: You want something like this to happen ?

image

Comment (Claiyc) [reply src/App.vue:351]: yes, I think this would actually be the better solution

Comment (BeierKevin) [reply src/App.vue:346]: OK I just commited the change

Comment (Claiyc) [reply src/App.vue:346]: just without the truncation tho

Comment (BeierKevin) [reply src/App.vue:346]: About the text truncing. I have already made a new issue #130 to make it possible to have different sizes of the notification and maybe icons etc.

And because we do not have any system using the notifications we still have to implement it into the other systems, and then we will maybe also run into things that we want to displayer or if we really need that much displayed text, like its a notification for the user and not a wikipedia page of some sorts

Comment (BeierKevin) [reply src/App.vue:346]: Without truncation and that the notification is currently set to a specific height it would look like this:

image

And tbh we didn't even decided even make a mock-up for how it should or could look maybe we should do something like this bevor we discuss on how we want it to look.

**Review by Claiyc — APPROVED** [review]
lgtm
